Инкремент №24

PaulYakow opened the pull request.

PaulYakow committed:
Спринт №10
+ Реализация инкремента №23 - конфигурация агента и сервера для отправки/получения запросов в доверенной подсети

PaulYakow committed:
+ Реализация инкремента №24 - изменение агента и сервера для отправки/получения gRPC-запросов

PaulYakow committed:
+ Исправление конфигурации агента

mr-fard4y commented on internal/app/client/client.go:
```diff
@@ -33,14 +33,15 @@ func Run(cfg *config.Config) {
 	go collector.Run(ctx, wg, cfg)
-	go sender.Run(ctx, wg, cfg)
+
+	if cfg.Address != "" {
```
сигнатуры запуска методов одинаковы
метод `Run` не должен содержать в себе логику инициализации поведения агента
необходимо лишь стартовать горутину

необходимо изменить методы `runHTTPSender` и `runGRPCSender`:
- логику инициализации агентов вынести в отдельные методы New 
- определить логику создания нужного типа агента - использовать фабричный метод; логику ветвления и выбора типа можно перенести в виде отдельного параметра в рамках конфига - clientType или похожий
- в зависимости от типа инициализировать агент, и вызывать метод `.Run` созданного типа агента

mr-fard4y commented on internal/controller/server/v1/middleware.go:
```diff
@@ -60,3 +61,40 @@ func (s *serverRoutes) decryptData(next http.Handler) http.Handler {
+				return
+			}
+
+			network, err := netip.ParsePrefix(trustedSubnet)
```
парсинг и преобразование можно перенести в рамках инициализации конфига или инициализации сервера, но не вызывать каждый раз при вызове хэндлера в рамках middlware

переменную `trustedSubnet` (или уже преообразованный вариант IP) можно хранить в структуре ` *serverRoutes`

mr-fard4y commented on internal/app/server/server.go:
```diff
@@ -72,6 +73,9 @@ func Run(cfg *config.Config) {
 		cfg.Address, cfg.StoreInterval, cfg.StoreFile, cfg.Restore, cfg.Key, cfg.Dsn, cfg.PathToCryptoKey)
 
+	// gRPC server
+	v2.New(serverUseCase, l, cfg)
```
тут лучше иметь возможность конфигрурировать тип запуска сервера:
- или только http;
- или только grpc;
- или в тандеме оба типа;

mr-fard4y replied:
по описанию в задании, можно воспользоваться подсказкой:
![image](https://user-images.githubusercontent.com/26343823/211362952-21eaa2db-60f6-4686-be4d-808116307070.png)

mr-fard4y commented on cmd/server/config/config.go:
```diff
@@ -35,6 +37,18 @@ var address = struct {
+	"grpc",
+	"g",
+	new(string),
+	"",
```
по желанию, можно задать некоторый специфичный порт для запуска, отличный от дефолтного

PaulYakow replied:
Намерено оставлены пустые значения, чтобы можно было опционально подключать gRPC

mr-fard4y commented on cmd/agent/config/config.go:
```diff
@@ -30,7 +32,19 @@ var address = struct {
+	"grpc",
+	"g",
+	new(string),
+	"",
```
по желанию, можно задать некоторый специфичный порт, отличный от дефолтного
в рамках дефолтной конфигурации сервера

PaulYakow replied:
Намерено оставлены пустые значения, чтобы можно было опционально подключать gRPC

mr-fard4y commented on internal/controller/client/sender_grpc.go:
```diff
@@ -0,0 +1,98 @@
+
+	request := &pb.UpdateBatchRequest{Metrics: grpcMetrics}
+
+	ctx, cancel := context.WithTimeout(ctx, 500*time.Millisecond)
```
переменную таймаут определить в рамках константы или переменной в рамках пакета/начале модуля

mr-fard4y commented on internal/controller/client/sender_grpc.go:
```diff
@@ -0,0 +1,98 @@
+		}
+	}
+
+	request := &pb.UpdateBatchRequest{Metrics: grpcMetrics}
```
реализовать аналогичную логику с хэдерами, как в http-клиенте
можно использоватье `metadata` из [стандартного пакета](https://pkg.go.dev/google.golang.org/grpc/metadata#pkg-overview)
[примеры использования](https://pkg.go.dev/google.golang.org/grpc/examples/features/metadata#section-readme)

PaulYakow committed:
Исправление замечаний по код-ревью №1:
+ controller/middleware -> парсинг доверенной сети перенесён в инициализацию контроллера, переменная добавлена в структуру
+ controller/sender_grpc -> переменная таймаута определена как константа
+ app/server -> добавлена структура сервера, выделена функция создания и метод запуска

PaulYakow committed:
+ app/server -> выделение создания различных типов репозиториев в методы, а также запуска планировщика

PaulYakow committed:
+ controller/sender_grpc -> логика работы с доверенной сетью (через metadata, аналогично заголовкам HTTP)
+ app/client -> разделена логика инициализации и запуска агентов (HTTP/gRPC)

mr-fard4y commented on internal/app/client/client.go:
```diff
@@ -18,29 +18,48 @@ import (
+	c.collector = client.NewCollector(c.usecase, c.logger)
 
-	agentUseCase := usecase.NewClientUC(ctx, agentRepo, agentHasher)
+	if cfg.Address != "" && cfg.GRPCTarget == "" {
```
текущую проверку лучше вынести методов в рамках конфига
`c.useHttpClient` или `c.useGRPCClient`, или аналогичные имена

mr-fard4y commented on internal/app/server/server.go:
```diff
@@ -19,73 +20,110 @@ import (
-		storage = true
-		l.Info("server - Run - PSQL in use")
+	// HTTP server
+	if cfg.Address != "" {
```
проверку можно вынести в рамках метода конфига
`c.useHttpService`

mr-fard4y commented on internal/app/server/server.go:
```diff
@@ -19,73 +20,110 @@ import (
+	}
+
+	// gRPC server
+	if cfg.GRPCAddress != "" {
```
проверку можно вынести в рамках метода конфига
`c.useGRPCService`

mr-fard4y commented on proto/metric.proto:
```diff
@@ -0,0 +1,64 @@
+message Metric {
+  string id = 1; // имя метрики
+  string mType = 2; // параметр, принимающий значение gauge или counter
+  int64 delta = 3; // значение метрики в случае передачи counter
```
`delta, value` можно указать [как `optional`](https://developers.google.com/protocol-buffers/docs/proto), т.к. в зависимости от типа метрики значение будет заполнено или нет

mr-fard4y reviewed (approved):
LGTM

PaulYakow committed:
+ Исправления по дополнениям к код-ревью №2
